### PR TITLE
Enabling full control of backdrop visibility from container

### PR DESCRIPTION
**Pull request by shiptoncraig**
I only just realised you've done a similar thing to what I was attempting - I've just made showBackdrop an two-way-bound property so we can hide/show it programmatically!

What do you think?

### CODE REVIEW

File: src/sidebar-container.component.ts [hunk 1]
```diff
@@ -5,9 +5,11 @@ import {
   Component,
   ContentChildren,
   Input,
+  Output,
```
Comment (arkon) [inline src/sidebar-container.component.ts]: Can we keep this sorted alphabetically?

File: src/sidebar.component.ts [hunk 2]
```diff
@@ -95,7 +95,7 @@ export class Sidebar implements OnChanges, OnDestroy {
   @Input() trapFocus: boolean = true;
   @Input() autoFocus: boolean = true;
 
-  @Input() showBackdrop: boolean = false;
+  @Input() showBackdropOnOpen: boolean = false;
```
Comment (arkon) [inline src/sidebar.component.ts]: I'd prefer not to rename this if possible to avoid breaking changes.

Comment (shiptoncraig) [inline src/sidebar.component.ts]: Will change it back

File: src/sidebar-container.component.ts (hunk 1 above)
Comment (shiptoncraig) [inline src/sidebar-container.component.ts]: Done

**Comment by shiptoncraig** [comment]
Do you prefer enableBackdrop from your commit or my (wordier) allowSidebarBackdropControl?

**Comment by arkon** [comment]
@shiptoncraig I think it depends on what functionality you're expecting. If `showBackdrop` is set to `true` regardless of a sidebar being open, would it just display the backdrop anyway?

File: src/sidebar-container.component.ts [hunk 3]
```diff
@@ -57,9 +59,10 @@ import { Sidebar } from './sidebar.component';
 })
 export class SidebarContainer implements AfterContentInit, OnDestroy {
   @Input() backdropClass: string;
+  @Input() allowSidebarBackdropControl: boolean = true;
 
-  /** @internal */
-  _showBackdrop: boolean = false;
+  @Input() showBackdrop: boolean = false;
+  @Output() showBackdropChange = new EventEmitter<null>();
```
Comment (arkon) [inline src/sidebar-container.component.ts]: I'm fairly confident that this should be `new EventEmitter<boolean>()`, and you need to actually emit it manually within `ngOnChange`.

File: README.md [hunk 4]
```diff
@@ -150,7 +150,14 @@ Various class names are attached to the sidebar and container for easier styling
 | Property name | Type | Default | Description |
 | ------------- | ---- | ------- | ----------- |
 | backdropClass | string | | Additional class name on the overlay element. |
+| showBackdrop | boolean | `false` | Controls the backdrop state of the sidebar container. This should be two-way bound. |
+| allowSidebarBackdropControl | boolean | `true` | Determines if the container component respects the sidebar's `showBackdropOnOpen` input. |
```
Comment (arkon) [inline README.md]: I guess this (long) name makes the most sense. Just make sure `showBackdropOnOpen` is changed to `showBackdrop`.

File: README.md [hunk 5]
```diff
@@ -167,7 +174,7 @@ Various class names are attached to the sidebar and container for easier styling
 | ariaLabel | string | | Value for the sidebar's `aria-label` attribute. |
 | trapFocus | boolean | `true` | Keeps focus within the sidebar when open. |
 | autoFocus | boolean | `true` | Automatically focus the first focusable element in the sidebar when opened. |
-| showBackdrop | boolean | `false` | If a translucent black backdrop overlay should appear over the page contents when the sidebar is open. |
+| showBackdrop | boolean | `false` | If a translucent black backdrop overlay should appear over the page contents when the sidebar is opened. |
```
Comment (arkon) [inline README.md]: Should note the container's precedence.

Comment (shiptoncraig) [inline README.md]: Good spot

File: src/sidebar-container.component.ts [hunk 6]
```diff
@@ -183,7 +194,7 @@ export class SidebarContainer implements AfterContentInit, OnDestroy {
         }
       }
 
-      this._showBackdrop = hasOpen;
+      this.showBackdrop = hasOpen;
```
Comment (arkon) [inline src/sidebar-container.component.ts]: You'll need to emit the change here as well.

File: README.md [hunk 7]
```diff
@@ -167,7 +174,7 @@ Various class names are attached to the sidebar and container for easier styling
 | ariaLabel | string | | Value for the sidebar's `aria-label` attribute. |
 | trapFocus | boolean | `true` | Keeps focus within the sidebar when open. |
 | autoFocus | boolean | `true` | Automatically focus the first focusable element in the sidebar when opened. |
-| showBackdrop | boolean | `false` | If a translucent black backdrop overlay should appear over the page contents when the sidebar is open. |
+| showBackdrop | boolean | `false` | If a translucent black backdrop overlay should appear over the page contents when the sidebar is opened.  This is ignored if the sidebar's parent container has its `allowSidebarBackdropControl` property set to true. |
```
Comment (arkon) [inline README.md]: Can you wrap that last `true` with backticks so it shows up as code?

**Comment by arkon** [comment]
Nice, thanks @shiptoncraig !

**Comment by arkon** [comment]
v3.3.0 published to NPM.

**Comment by shiptoncraig** [comment]
Great, thanks!

**Pull request merged**